### ISSUE TITLE
User should be able to view the effective query plan for a query

### ISSUE THREAD

**Issue by hmottestad** [eclipse/rdf4j#2044]
It is useful for the user to be able to inspect the effective query plan for a query. There should be some way for a user to view the query plan.

The query plan is usually not finalised until after the query has been executed, but before the first result is retrieved.

The query plan generated when the query is instantiated is not the optimised plan that actually gets run.

Options for exposing the query plan:
 - String explain() method on QueryResult
 - String explain(QueryResult qr) on QueryResults

It would also be useful if `StatementPattern` showed the estimated cardinality.

**Comment by hmottestad** [comment]
## Example query and plan

### Query
```
SELECT ?language (count(?a) as ?count) where {
        ?b dcat:dataset ?a .
        ?a dct:language ?language .
        FILTER(?language != eu-lang:ENG)
} group by ?language

````
### Plan
```
Projection
   ProjectionElemList
      ProjectionElem "language"
      ProjectionElem "count"
   Extension
      ExtensionElem (count)
         Count
            Var (name=a)
      Group (language)
         Join
            Filter
               Compare (!=)
                  Var (name=language)
                  ValueConstant (value=http://publications.europa.eu/resource/authority/language/ENG)
               StatementPattern
                  Var (name=a)
                  Var (name=_const_2783f1e8_uri, value=http://purl.org/dc/terms/language, anonymous)
                  Var (name=language)
            StatementPattern
               Var (name=b)
               Var (name=_const_7420303d_uri, value=http://www.w3.org/ns/dcat#dataset, anonymous)
               Var (name=a)
         GroupElem
            Count
               Var (name=a)
```

**Comment by hmottestad** [comment]
@jeenbroekstra do you think this is a good idea?

**Comment by kenwenzel** [comment]
Maybe you could also use some kind of magic keyword like `EXPLAIN` in front of the actual SPARQL query as in:
```
EXPLAIN
SELECT ?language (count(?a) as ?count) where {
        ?b dcat:dataset ?a .
        ?a dct:language ?language .
        FILTER(?language != eu-lang:ENG)
} group by ?language
```
The result could be a JSON object that represents the query plan.
This would also work with remote servers that support the EXPLAIN feature.

**Comment by kenwenzel** [comment]
Oracle provides also the possibility to fetch the last query plan via SQL:
[https://stackoverflow.com/questions/4376329/jdbc-oracle-fetch-explain-plan-for-query](https://stackoverflow.com/questions/4376329/jdbc-oracle-fetch-explain-plan-for-query)

**Comment by jeenbroekstra** [comment]
It would definitely be useful to make this more accessible, but a couple of caveats:

1. We're not a single database, we're a framework for different databases. Not every RDF4J store implementation will have the option to provide a query plan in terms of the RDF4J algebra (or even at all). That means that we can't simply standardize on the string representation of our query algebra as the mandated output format, and we can't rely on any sort of report always being present.
2. Serializing the query plan as part of every query execution is likely to be a significant performance burden. We should look into a way to minimize this, so that it's only done when the user actually needs/wants it.

On point 1: we can of course provide a generic mechanism with a specific implementation for our out of the box stores, but then make it such that third party implementors can use it with minimal asssumptions about what their internal query implementation looks like. 

On point 2 I'd be keen to go with something similar to @kenwenzel 's suggestion of a separate 'explain' function (either as an API call or as an extension to the language, haven't thought that through yet), so that we can completely decouple this functionality from the actual execution of the query.

It might be worth looking into what commercial triplestores like GraphDB or Stardog have in this space, as well.

**Comment by jeenbroekstra** [comment]
As an aside: I've long thought that we should look into a more up-to-date string representation of our algebra tree, one that perhaps is a bit more aligned with the terminology in the SPARQL recommendation. The current names are really pre-SPARQL (the algebra tree was originally designed for the SeRQL language).

**Comment by hmottestad** [comment]
#### Stardog
 - visual query plans: https://www.stardog.com/blog/visual-query-plans-in-studio/
 - CLI command: https://www.stardog.com/docs/man/query-explain
 - Java API with .explain method on query object: https://www.stardog.com/docs/java/snarl/com/complexible/stardog/api/query#explain--

#### PostgreSQL
 - query plan as json: http://www.postgresonline.com/journal/archives/171-Explain-Plans-PostgreSQL-9.0-Text,-JSON,-XML,-YAML-Part-1-You-Choose.html

#### Jena
 - logging based approach: https://jena.apache.org/documentation/query/explain.html

#### GraphDB
 - pesudo-graph approach: http://graphdb.ontotext.com/documentation/standard/explain-plan.html

#### ShaclSail
 - configurable logging of validation (query) plans as graphviz dot: [link](https://dreampuf.github.io/GraphvizOnline/#digraph%20%20%7B%0Alabelloc%3Dt%3B%0Afontsize%3D30%3B%0Alabel%3D%22DatatypePropertyShape%22%3B%0A1866229258%20%5Blabel%3D%22Base%20sail%22%20nodeShape%3Dpentagon%20fillcolor%3Dlightblue%20style%3Dfilled%5D%3B%0A1555990397%20%5Blabel%3D%22Added%20statements%22%20nodeShape%3Dpentagon%20fillcolor%3Dlightblue%20style%3Dfilled%5D%3B%0A1544078442%20%5Blabel%3D%22Removed%20statements%22%20nodeShape%3Dpentagon%20fillcolor%3Dlightblue%20style%3Dfilled%5D%3B%0A1337866219%20%5Blabel%3D%22Previous%20state%20connection%22%20nodeShape%3Dpentagon%20fillcolor%3Dlightblue%20style%3Dfilled%5D%3B%0A1291367132%20%5Blabel%3D%22DirectTupleFromFilter%22%5D%3B%0A1887699190%20%5Blabel%3D%22DatatypeFilter%7Bdatatype%3Dhttp%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23integer%7D%22%5D%3B%0A1479140596%20-%3E%201887699190%0A1887699190%20-%3E%201291367132%20%5Blabel%3D%22false%20values%22%5D%0A1479140596%20%5Blabel%3D%22UnionNode%22%5D%3B%0A1108889615%20-%3E%201479140596%0A1108889615%20%5Blabel%3D%22UnionNode%22%5D%3B%0A1275028674%20-%3E%201108889615%0A455888635%20%5Blabel%3D%22BufferedSplitter%22%5D%3B%0A204805934%20-%3E%20455888635%0A204805934%20%5Blabel%3D%22TrimTuple%7BnewLength%3D1%7D%22%5D%3B%0A204322447%20-%3E%20204805934%0A204322447%20%5Blabel%3D%22Select%7Bquery%3D’select%20*%20where%20%7B%20BIND(rdf%3Atype%20as%20%3Fb)%20%5Cn%20BIND(%3Chttp%3A%2F%2Fexample.com%2Fns%23Person%3E%20as%20%3Fc)%20%5Cn%20%3Fa%20%3Fb%20%3Fc.%7D%20order%20by%20%3Fa'%7D%22%5D%3B%0A1555990397%20-%3E%20204322447%0A1275028674%20%5Blabel%3D%22InnerJoin%22%5D%3B%0A455888635%20-%3E%201275028674%20%5Blabel%3D%22left%22%5D%3B%0A1019484860%20-%3E%201275028674%20%5Blabel%3D%22right%22%5D%3B%0A1019484860%20%5Blabel%3D%22DirectTupleFromFilter%22%5D%3B%0A1164365897%20%5Blabel%3D%22DatatypeFilter%7Bdatatype%3Dhttp%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23integer%7D%22%5D%3B%0A1640899500%20-%3E%201164365897%0A1164365897%20-%3E%201019484860%20%5Blabel%3D%22false%20values%22%5D%0A1640899500%20%5Blabel%3D%22Select%7Bquery%3D’select%20*%20where%20%7B%20%3Fa%20%3Chttp%3A%2F%2Fexample.com%2Fns%23age%3E%20%3Fc.%20%7D%20order%20by%20%3Fa'%7D%22%5D%3B%0A1555990397%20-%3E%201640899500%0A1275028674%20-%3E%203565780%20%5Blabel%3D%22discardedRight%22%5D%3B%0A473666452%20-%3E%201108889615%0A473666452%20%5Blabel%3D%22ExternalTypeFilterNode%7BfilterOnType%3Dhttp%3A%2F%2Fexample.com%2Fns%23Person%7D%22%5D%3B%0A3565780%20-%3E%20473666452%0A1337866219%20-%3E%20473666452%20%5Blabel%3D%22filter%20source%22%5D%0A3565780%20%5Blabel%3D%22BufferedTupleFromFilter%22%5D%3B%0A1865219266%20-%3E%201479140596%0A1865219266%20%5Blabel%3D%22BulkedExternalInnerJoin%7Bpredicate%3Dnull%2C%20query%3D'%3Fa%20%3Chttp%3A%2F%2Fexample.com%2Fns%23age%3E%20%3Fc.%20'%7D%22%5D%3B%0A455888635%20-%3E%201865219266%20%5Blabel%3D%22left%22%5D%0A1337866219%20-%3E%201865219266%20%5Blabel%3D%22right%22%5D%0A%7D%0A8)

I would like to recommend that we consider the following:
 - discoverability: a user/developer should be able to discover the explain functionality if they already understand RDF4J without needing to google or read documentation
 - output should be easy to read
 - output should be configurable so that we can support visual query plans
 - our internal format should be an open format that is easy to extend - i prefer json

**Comment by Akash-Sharma-1** [comment]
Hi! Can I work on this issue? Seems Interesting to me. I would like to communicate with the maintainers. Thanks

**Comment by jeenbroekstra** [comment]
Hi @Akash-Sharma-1, we welcome contributions of course, but I should point that this issue is currently already under active development by @hmottestad , so it's perhaps not the best starting place. Then again, if you have specific feedback or ideas for tackling this issue that you wish to discuss, we'd of course welcome it!

**Comment by hmottestad** [comment]
@Akash-Sharma-1 I'm not against help on this issue, but the issue in itself still needs a few more weeks of discussion I think before we can move along with anything major. We still need to decide on the following:
 - format for query plan that can be generated locally or transmitted over the wire
 - how the api should look
 - how the user can use it from the workbench
 - how the user can use it from a SPARQL endpoint
 - coordinate with vendors (especially Ontotext)

**Comment by jeenbroekstra** [comment]
This is rather a large piece of work. Can we split this up into multiple separate subtasks?

**Comment by aschwarte10** [comment]
I haven't follow the changes deeply yet, but one comment from my side

@hmottestad for additional consideration and future alignment consider how I expose the query plan feature in the FedX federation, see `org.eclipse.rdf4j.federated.QueryManager.getQueryPlan(String)`. This method does not do any evaluation of the query, but just returns the query plan, sort of an API. Maybe something like this can be generalized

Besides that one can obviously increase logging levels.

**Comment by hmottestad** [comment]
I took a look at that method. It just prints our internal query plan. What I initially thought I would do too. I have come to realise though that maybe we don't want to expose it like that in a public and documented way, since it makes it hard to maintain backwards compatibility in the future if someone tries to parse the format.

**Comment by hmottestad** [comment]
I would like to propose the following API:

```java
try (SailRepositoryConnection connection1 = repository.getConnection()) {
	TupleQuery tupleQuery = connection1.prepareTupleQuery("select * where {?a ?b ?c}");

	//Option 1
	String queryPlan1 = tupleQuery.getQueryPlan(QueryPlan.Unoptimized, QueryPlanFormat.Humanreable);
	String queryPlan2 = tupleQuery.getQueryPlan(QueryPlan.Unoptimized, QueryPlanFormat.Json);
	String queryPlan3 = tupleQuery.getQueryPlan(QueryPlan.Optimized, QueryPlanFormat.Json);
	String queryPlan4 = tupleQuery.getQueryPlan(QueryPlan.Executed, QueryPlanFormat.Json);

	//Option 2 allows the return type to be more generic
	String queryPlan5 = tupleQuery.getQueryPlan(QueryPlan.Unoptimized).as(QueryPlanFormat.Humanreable);
	Map<String, Object> queryPlan6 = tupleQuery.getQueryPlan(QueryPlan.Unoptimized).as(QueryPlanFormat.JsonMap);
	String queryPlan7 = tupleQuery.getQueryPlan(QueryPlan.Unoptimized).as(QueryPlanFormat.JsonString);
	TupleExpr queryPlan8 = tupleQuery.getQueryPlan(QueryPlan.Unoptimized).as(QueryPlanFormat.TupleExpr);

}
```

For the server I propose to extend the current query implementation to support new formats specified as an Accept header from the client. For instance: `Accept: application/queryplan+json`. Together with a new param "queryPlan" (or similar)  to specify unoptimized/optimized/executed ("optimized" being the default).

Then a checkbox and a dropdown in the workbench to interact with this.

What does everyone think?

**Comment by jeenbroekstra** [comment]
Good overview. I think I prefer option 2's output format approach - gives us a lot of flexibility to return a generic QueryPlan object that in the default implementation can just be empty. 

Slight preference to use `explain` as the method name since it's more succinct yet still accurately captures the intent. 

As for whether to go with an addition to the `Query` interface vs a utility method: I don't agree that this is somehow "non-OO", it's perfectly valid to treat functions as first class citizens in an object-oriented language. We deliberately do this in multiple places in the RDF4J framework, and it's often a good choice because it reduces coupling. 

Having said that I believe a static function like this should be free from side effects, and  since in this case creating the plan will have a side effect (namely that it will either optimize or completely execute the query, and therefore change the structure of the input Query object), I don't think using a util function is a good choice here.  

> For the server I propose to extend the current query implementation to support new formats specified as an Accept header from the client. For instance: Accept: application/queryplan+json. 

Do we need a custom MIME-type for this? Why not just have a separate resource endpoint to GET a query's explanation, and have it return either `text/plain`, `application/json`, or even some RDF serialization?

**Comment by hmottestad** [comment]
We now have two things left:
 - remote repo against RDF4J Server
 - fedx?

**Comment by jeenbroekstra** [comment]
Closing this, created followups for the things you mentioned being left.
